google-cloud-storage/src/test/java/com/google/cloud/storage/it/runner/registry/StorageInstance.java

```diff
@@ -103,9 +67,119 @@ private static VetoedBucketUpdateException err(String bucketName) {
     return new VetoedBucketUpdateException("Attempted to modify global bucket: " + bucketName);
   }
 
-  private static final class VetoedBucketUpdateException extends RuntimeException {
+  private static class VetoException extends RuntimeException {
+    private VetoException(String message) {
+      super(message);
+    }
+  }
+
+  private static final class VetoedBucketUpdateException extends VetoException {
     private VetoedBucketUpdateException(String message) {
       super(message);
     }
   }
+
+  /**
+   * Define a proxy which can veto calls attempting to mutate protected buckets. this helps guard
+   * against a test trying to mutate the global bucket rather than creating its own bucket.
+   */
+  private final class VetoingStorageProxy extends AbstractStorageProxy {
+
+    private VetoingStorageProxy() {
+      super(storage);
+    }
+
+    @Override
+    public Bucket update(BucketInfo bucketInfo, BucketTargetOption... options) {
+      checkBucketProtected(bucketInfo);
+      return super.update(bucketInfo, options);
+    }
+
+    @Override
+    public boolean delete(String bucket, BucketSourceOption... options) {
+      checkBucketProtected(bucket);
+      return super.delete(bucket, options);
+    }
+
+    @Override
+    public Policy setIamPolicy(String bucket, Policy policy, BucketSourceOption... options) {
+      checkBucketProtected(bucket);
+      return super.setIamPolicy(bucket, policy, options);
+    }
+
+    @Override
+    public boolean deleteDefaultAcl(String bucket, Entity entity) {
+      checkBucketProtected(bucket);
+      return super.deleteDefaultAcl(bucket, entity);
+    }
+
+    @Override
+    public Acl createDefaultAcl(String bucket, Acl acl) {
+      checkBucketProtected(bucket);
+      return super.createDefaultAcl(bucket, acl);
+    }
+
+    @Override
+    public Acl updateDefaultAcl(String bucket, Acl acl) {
+      checkBucketProtected(bucket);
+      return super.updateDefaultAcl(bucket, acl);
+    }
+
+    @Override
+    public boolean deleteAcl(String bucket, Entity entity, BucketSourceOption... options) {
+      checkBucketProtected(bucket);
+      return super.deleteAcl(bucket, entity, options);
+    }
+
+    @Override
+    public boolean deleteAcl(String bucket, Entity entity) {
```

Comment:
By having this protection, I'm concerned about future changes not being caught by this mechanism. However, IIUC, whenever new methods are added to the Storage interface, the new methods will need to be implemented in the Vetoing class as well so it covers this case.

Reply:
Correct, by explicitly defining it as an implementation of the interface we will have to implement the method if a new one is added, whereas the runtime generated proxy before was using reflection to selectively choose which methods to augment.